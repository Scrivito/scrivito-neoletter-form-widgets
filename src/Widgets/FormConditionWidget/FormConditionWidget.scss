.scrivito-neoletter-form-widgets .condition-info {
  position: absolute;
  top: 6px;
  left: 50%;
  background: white;
  transform: translate(-50%);
  padding: 0 5px;
  margin: 0;
}
