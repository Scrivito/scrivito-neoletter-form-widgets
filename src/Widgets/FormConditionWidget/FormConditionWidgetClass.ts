import * as Scrivito from "scrivito";

export const FormConditionWidget = Scrivito.provideWidgetClass(
  "FormConditionWidget",
  {
    onlyInside: "FormConditionalContainerWidget",
    attributes: {
      title: "string",
      content: "widgetlist"
    }
  }
);
