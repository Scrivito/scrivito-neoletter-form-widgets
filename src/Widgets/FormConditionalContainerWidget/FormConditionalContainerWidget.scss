.scrivito-neoletter-form-widgets {
  .condition-wrapper {
    border: green 2px dotted;
    .select-title {
      margin-top: 0.3em;
      p {
        display: inline;
      }
    }
  }
  .conditional-content {
    border: orange 2px dotted;
  }
  .header-info {
    position: absolute;
    top: 6px;
    left: 50%;
    background: white;
    transform: translate(-50%);
    padding: 0 5px;
    margin: 0;
  }
  .dropdown-container {
    select {
      width: 100%;
    }
  }
}
