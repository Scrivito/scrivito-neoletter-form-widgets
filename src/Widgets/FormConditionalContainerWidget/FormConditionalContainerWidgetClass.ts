import * as Scrivito from "scrivito";

export const FormConditionalContainerWidget = Scrivito.provideWidgetClass(
  "FormConditionalContainerWidget",
  {
    attributes: {
      headerType: [
        "enum",
        {
          values: ["dropdown", "radio"]
        }
      ],
      title: "string",
      customFieldName: "string",
      conditions: ["widgetlist", { only: "FormConditionWidget" }],
      required: "boolean",
      helpText: "html"
    }
  }
);
