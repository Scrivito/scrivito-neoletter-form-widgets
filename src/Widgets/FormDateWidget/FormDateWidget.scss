.scrivito-neoletter-form-widgets .form-date {
  .date-title {
    p {
      display: inline;
    }
  }
}
