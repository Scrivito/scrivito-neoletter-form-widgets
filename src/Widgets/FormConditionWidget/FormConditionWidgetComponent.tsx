import * as React from "react";
import * as Scrivito from "scrivito";
import { InPlaceEditingPlaceholder } from "../../Components/InPlaceEditingPlaceholder";
import { FormConditionWidget } from "./FormConditionWidgetClass";
import "./FormConditionWidget.scss";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
Scrivito.provideComponent(FormConditionWidget, ({ widget, getData }: any) => {
  const data = getData ? getData(widget.id()) : { isActive: false };

  if (!data.isActive && !Scrivito.isInPlaceEditingActive()) {
    return null;
  }

  if (!widget.get("content").length) {
    return (
      <InPlaceEditingPlaceholder center>
        Select widgets in the widget properties.
      </InPlaceEditingPlaceholder>
    );
  }

  return (
    <>
      {Scrivito.isInPlaceEditingActive() && (
        <span className="condition-info">
          {"Condition: " + widget.get("title")}
        </span>
      )}
      {widget.get("content") && widget.get("content").length > 0 && (
        <Scrivito.ContentTag
          content={widget}
          attribute="content"
          className={
            Scrivito.isInPlaceEditingActive() ? "conditional-content" : ""
          }
        />
      )}
    </>
  );
});
