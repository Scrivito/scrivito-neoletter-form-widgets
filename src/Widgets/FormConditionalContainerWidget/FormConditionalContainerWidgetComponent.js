import * as React from "react";
import * as Scrivito from "scrivito";
import "./FormConditionalContainerWidget.scss";
import { ConditionalHeader } from "../FormContainerWidget/components/ConditionalHeaderComponent";

Scrivito.provideComponent("FormConditionalContainerWidget", ({ widget }) => {
  const [selectedConditionId, setSelectedConditionId] = React.useState("");
  const isDropdownHeader = widget.get("headerType") == "dropdown";
  const onChangeSelected = (e) => {
    const selectedId = isDropdownHeader
      ? e.target.options[e.target.selectedIndex].id
      : e.target.id;
    setSelectedConditionId(selectedId);
  };

  return (
    <>
      <ConditionalHeader widget={widget} onChangeSelected={onChangeSelected} />
      <Scrivito.ContentTag
        content={widget}
        attribute="conditions"
        widgetProps={{
          getData: (conditionId) => {
            const conditions = widget.get("conditions");
            let isActive = false;
            conditions.some((condition) => {
              if (condition.id() == conditionId) {
                isActive = selectedConditionId == conditionId;
                return true;
              }
            });
            return { isActive };
          },
        }}
      />
    </>
  );
});
