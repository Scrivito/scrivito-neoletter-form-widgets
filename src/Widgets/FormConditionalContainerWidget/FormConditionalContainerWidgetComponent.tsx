import * as React from "react";
import * as Scrivito from "scrivito";
import { ConditionalHeader } from "../FormStepContainerWidget/components/ConditionalHeaderComponent";
import { FormConditionalContainerWidget } from "./FormConditionalContainerWidgetClass";
import { StringMap } from "../../../types/types";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { isEmpty } from "../FormStepContainerWidget/utils/lodashPolyfills";
import "./FormConditionalContainerWidget.scss";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
Scrivito.provideComponent(FormConditionalContainerWidget, ({ widget, onInputChange }: any) => {
  const [selectedConditionId, setSelectedConditionId] = React.useState("");
  const isDropdownHeader = widget.get("headerType") == "dropdown";

  const resetFieldsInConditions = (conditions: Scrivito.Widget[]) => {
    const resetFields = conditions
      .flatMap(condition => condition.get("content") as Scrivito.Widget[])
      .map(widget => getFieldName(widget))
      .filter(fieldName => !isEmpty(fieldName))
      .reduce((acc, fieldName) => {
        acc[fieldName] = "";
        return acc;
      }, {} as StringMap<string>);

    onInputChange(resetFields);
  };

  const onChangeSelected = (e: React.BaseSyntheticEvent) => {
    const selectedId = isDropdownHeader
      ? e.target.options[e.target.selectedIndex].id
      : e.target.id;
    setSelectedConditionId(selectedId);
    if (!onInputChange) {
      return;
    }
    // update the field corresponding to the selected condition
    onInputChange(widget.get("customFieldName"), e.target.value);

    // gather all field names to reset for the current condition
    const conditions: Scrivito.Widget[] = widget.get("conditions");
    resetFieldsInConditions(conditions);
  };

  return (
    <>
      <ConditionalHeader widget={widget} onChangeSelected={onChangeSelected} />
      <Scrivito.ContentTag
        content={widget}
        attribute="conditions"
        widgetProps={{
          getData: (conditionId: string) => {
            const conditions: Scrivito.Widget[] = widget.get("conditions");
            let isActive = false;
            conditions.some(condition => {
              if (condition.id() == conditionId) {
                isActive = selectedConditionId == conditionId;
                return true;
              }
            });
            return { isActive };
          },
          onInputChange
        }
        }
      />
    </>
  );
});
