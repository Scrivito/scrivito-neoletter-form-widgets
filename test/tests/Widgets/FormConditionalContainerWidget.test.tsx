import * as Scrivito from "scrivito";
import { FormConditionWidget } from "../../../src/Widgets/FormConditionWidget/FormConditionWidgetClass";
import PageRenderer from "../../helpers/pageRenderer";
import { FormDateWidget } from "../../../src/Widgets/FormDateWidget/FormDateWidgetClass";
import { FormConditionalContainerWidget } from "../../../src/Widgets/FormConditionalContainerWidget/FormConditionalContainerWidgetClass";
import "../../../src/Widgets/FormConditionalContainerWidget/FormConditionalContainerWidgetComponent";
import "../../../src/Widgets/FormConditionWidget/FormConditionWidgetComponent";
import "../../../src/Widgets/FormDateWidget/FormDateWidgetComponent";
import { fireEvent } from "@testing-library/react";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();
const conditions = [
  new FormConditionWidget({
    title: "select me",
    content: [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ]
  }),
  new FormConditionWidget({
    title: "i am better :)",
    content: [
      new FormDateWidget({
        customFieldName: "custom_date2",
        title: "date",
        dateType: "date"
      })
    ]
  })
];
const widgetProps = {
  headerType: "radio",
  title: "Select something",
  customFieldName: "custom_test",
  conditions: conditions,
  required: false,
  helpText: ""
};

describe("FormConditionalContainerWidget", () => {
  it("should show header-info & condition-info if isInPlaceEditingActive is true", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(true);
    pageRenderer.render({
      body: [new FormConditionalContainerWidget(widgetProps)]
    });
    const condtionalContentInfo = document.querySelector(
      ".conditional-content"
    );
    const condtionalHeaderInfo = document.querySelector(".header-info");
    expect(condtionalContentInfo).toBeInTheDocument();
    expect(condtionalHeaderInfo).toBeInTheDocument();
  });

  it("should have radios as header", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(false);

    pageRenderer.render({
      body: [
        new FormConditionalContainerWidget({
          ...widgetProps,
          conditions: [
            // reassign is necessary, otherwise conditions will be empty
            new FormConditionWidget({
              title: "select me",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date",
                  title: "date",
                  dateType: "date"
                })
              ]
            }),
            new FormConditionWidget({
              title: "i am better :)",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date2",
                  title: "date",
                  dateType: "date"
                })
              ]
            })
          ]
        })
      ]
    });

    const headerInputs = document.querySelectorAll(".form-check-input");
    expect(headerInputs).toHaveLength(2);
    headerInputs.forEach(input => {
      expect(input).toHaveAttribute("type", "radio");
    });
  });
  it("should have dropdown as header", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(false);

    pageRenderer.render({
      body: [
        new FormConditionalContainerWidget({
          ...widgetProps,
          headerType: "dropdown",
          conditions: [
            // reassign is necessary, otherwise conditions will be empty
            new FormConditionWidget({
              title: "select me",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date",
                  title: "date",
                  dateType: "date"
                })
              ]
            }),
            new FormConditionWidget({
              title: "i am better :)",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date2",
                  title: "date",
                  dateType: "date"
                })
              ]
            })
          ]
        })
      ]
    });
    const select = document.querySelector("select");
    const dropdownOptions = document.querySelectorAll("option");
    expect(select).toBeInTheDocument();
    expect(dropdownOptions).toHaveLength(3);
  });

  it("should select the second option in the dropdown and show the corresponding condition for it", () => {
    jest.spyOn(Scrivito, "isInPlaceEditingActive").mockReturnValue(false);

    const { container } = pageRenderer.render({
      body: [
        new FormConditionalContainerWidget({
          ...widgetProps,
          headerType: "dropdown",
          conditions: [
            new FormConditionWidget({
              title: "select me",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date_first_condition",
                  title: "date",
                  dateType: "date"
                })
              ]
            }),
            new FormConditionWidget({
              title: "i am better :)",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date_second_condtion",
                  title: "date",
                  dateType: "date"
                })
              ]
            })
          ]
        })
      ]
    });

    const select = container.querySelector("select")!;
    // Trigger a change event with the second option as selected
    select.value = "i am better :)";
    fireEvent.change(select);

    expect(select.value).toBe("i am better :)");
    const unSelectedCondition = container.querySelector(
      'input[name="custom_date_first_condtion"]'
    );
    const selectedCondition = container.querySelector(
      'input[name="custom_date_second_condtion"]'
    );
    expect(unSelectedCondition).not.toBeInTheDocument();
    expect(selectedCondition).toBeInTheDocument();
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [
        new FormConditionalContainerWidget({
          ...widgetProps,
          conditions: [
            // reassign is necessary, otherwise conditions will be empty
            new FormConditionWidget({
              title: "select me",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date",
                  title: "date",
                  dateType: "date"
                })
              ]
            }),
            new FormConditionWidget({
              title: "i am better :)",
              content: [
                new FormDateWidget({
                  customFieldName: "custom_date2",
                  title: "date",
                  dateType: "date"
                })
              ]
            })
          ]
        })
      ]
    });
    expect(tree).toMatchSnapshot();
  });
});
