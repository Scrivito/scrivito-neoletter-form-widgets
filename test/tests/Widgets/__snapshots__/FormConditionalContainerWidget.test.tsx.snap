// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormConditionalContainerWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="mb-3 select-container "
    >
      <div
        className="select-title"
      >
        <span
          className="text-super"
        >
          Select something
        </span>
      </div>
      <div
        className="row "
      >
        <label
          className="select-label "
        >
          <input
            className="form-check-input"
            id="staticWidgetID"
            name="custom_test"
            onChange={[Function]}
            required={false}
            type="radio"
            value="select me"
          />
          <span>
            select me
          </span>
        </label>
        <label
          className="select-label "
        >
          <input
            className="form-check-input"
            id="staticWidgetID"
            name="custom_test"
            onChange={[Function]}
            required={false}
            type="radio"
            value="i am better :)"
          />
          <span>
            i am better :)
          </span>
        </label>
      </div>
    </div>
    <div>
      <div />
      <div />
    </div>
  </div>
</div>
`;
